src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -537,18 +539,42 @@ void MoveReturnEndToNextBlock()
                     _readerAwaitable.SetUncompleted();
                 }
 
-                while (returnStart != null && returnStart != returnEnd)
+                success = _operationState.TryEndRead();
+            }
+
+            if (!success)
+            {
+                TrySchedule(_writerScheduler, completionData);
+            }
+
+            if (returnStart != null && returnStart != returnEnd)
+            {
+                // Reset and recycle the memory blocks outside the lock;
+                // the Pool has its own synchronization and resetting the memory is slower.
+                BufferSegment? next = returnStart;
+                do
                 {
-                    BufferSegment? next = returnStart.NextSegment;
-                    returnStart.ResetMemory();
-                    ReturnSegmentUnsynchronized(returnStart);
-                    returnStart = next;
-                }
+                    next.ResetMemory();
+                    next = next.NextSegment;
+                } while (next != null && next != returnEnd);
 
-                _operationState.EndRead();
+                // Return the segments inside the lock
+                lock (_sync)
```

Comment:
Would it be better to have a distinct lock object for accessing the _bufferSegmentPool?

You would have to take both `_sync` and the new lock object to get a new `_writingHead`, but returning would only need to take the new one.

Reply:
Done